### ISSUE TITLE
Loader usage with preact-cli

### ISSUE THREAD

**Issue by cherta** [smyte/jsxstyle#98]
Hi, I'm not sure this is a `jsxstyle` issue but I think it worth putting it here since the library has built-in support for preact and preact-cli is widely used.

I'm working on a `preact-cli` app that heavily uses `jsxstyle-preact` and the [pre-render](https://github.com/developit/preact-cli#pre-rendering) feature.

For that, I tried to configure the loader using the `preact.config.js` file which allows you to mess with the webpack configuration before building the project.

Here's the snippet:

```js
import JsxstyleLoaderPlugin from 'jsxstyle-loader/plugin';
export default function (config, env, helpers) {
	if (env.production) {
		config.plugins.push(new JsxstyleLoaderPlugin());

		let { index } = helpers.getLoadersByName(config, 'babel-loader')[0];
		config.module.loaders.splice(index + 1, 0, { test: /\.js$/, loader: 'jsxstyle-loader' });
	}
}
```

What this function does is to place the `jsxstyle-loader` _after_ the babel one and before the `css-loader`.

The webpack config generated looks a bit like this one:

```js
module.exports = {
  plugins: [
    new JsxstyleLoaderPlugin()
  ],
  //...
  module: {
    loaders: [
      {
        enforce: "pre",
        test: /\.jsx?$/,
        loader: "babel-loader",
        options: { babelrc: false, presets: [/*removed for clarity*/], plugins: [/*removed for clarity*/] }
      },
      { test: /\.js$/, loader: "jsxstyle-loader" },
      //...
      {
        test: /\.(css|less|s[ac]ss|styl)$/,
        include: [
          "/Users/cherta/workspace/ingsw/website/src/components",
          "/Users/cherta/workspace/ingsw/website/src/routes"
        ],
        loader: [
          { loader: '/Users/cherta/workspace/ingsw/website/node_modules/extract-text-webpack-plugin/dist/loader.js',
       options: [/*removed for clarity*/] },
          { loader: 'style-loader' },
          { loader: 'css-loader', options: [/*removed for clarity*/] },
          { loader: 'postcss-loader', options: [/*removed for clarity*/] }
        ]
      },
      //...
    ]
  }
}
```

So far I couldn't make it work and styles are generated after the JS loads. I really appreciate any help since I'm not a webpack expert and these things always seem tricky.

**Comment by meyer** [comment]
From a cursory look I can tell you that `jsxstyle-loader` should run _before_ `babel-loader`. It’s in the right position in the config (below `babel-loader`) but `babel-loader` runs first since it’s a preloader. Potential workaround is to specify `enforce: "pre"` on `jsxstyle-loader` as well, but that might cause issues.

I’ll give `jsxstyle-loader` a try with `preact-cli` here in a bit and let you know how it goes. It’d probably be a good example to add to [the examples folder](https://github.com/smyte/jsxstyle/tree/610d5eb/examples).

**Comment by meyer** [comment]
@cherta I’ve just about got a working example going with `preact-cli`. One issue I’m running into is that babel seems to be prepending `import { h } from "preact";` to the top of any file that contains JSX, regardless of whether or not that `h` import already exists in the file.

Example transpiled source:

```jsx
import { h } from "preact";

function _classCallCheck(instance, Constructor) { /* truncated */ }

function _possibleConstructorReturn(self, call) { /* truncated */ }

function _inherits(subClass, superClass) { /* truncated */ }

/** @jsx h */ //  <-- the start of the transpiled file
import { h, Component } from "preact";
import { Router } from "preact-router";
import { Block } from "jsxstyle/preact";

import Header from "./Header";
// etc.
```

Still digging…

**Comment by meyer** [comment]
`babel-plugin-jsx-pragmatic` seems to be the culprit here. It prepends `import { h } from "preact";` to every JS file automatically. Apparently `preact-cli` was always adding a duplicate import, but babel only checks for duplicate imports during scope extraction—a feature `jsxstyle-loader` makes heavy use of.

I ran into another problem. Apparently `preact-cli` defaults to using `css-loader` with CSS modules enabled, which means that class names in CSS files get transformed into _hashed_ class names, causing a class name mismatch between generated CSS and generated JS.

The solution on the `preact-cli` side would require some pretty nasty hacks to the generated webpack config and the result would be incredibly brittle. This CSS modules issue is likely to pop up again, so I added a (undocumented for now) `cssModules` loader option to `jsxstyle-loader` that should mitigate the issue. Basically, it adds a `:global` prefix to class names generated by `jsxstyle-loader`. This will ensure that the class names don’t get mangled down the line by a loader that supports CSS modules.

You can see the change here: https://github.com/smyte/jsxstyle/commit/9b416fc#diff-b29744433a39bcec0b5c6359bb3e0b9dR935

You can see the `preact-cli` example here: [examples/preact-cli](https://github.com/smyte/jsxstyle/tree/9b416fc/examples/preact-cli)

The `cssModules` option will be available as soon as I publish to NPM.

Do let me know if you have any questions.

**Comment by meyer** [comment]
Published as `jsxstyle-loader@1.0.11`. I’m going to close this for now, but feel free to reopen if the `preact-cli` example I added doesn’t solve your issue.
